specs/core/0_beacon-chain.md

```diff
@@ -1264,50 +1262,59 @@ def get_genesis_beacon_state(genesis_validator_deposits: List[Deposit],
 
 ## Beacon chain state transition function
 
-We now define the state transition function. At a high level, the state transition is made up of four parts:
+The post-state corresponding to a pre-state `state` and a block `block` is defined as `state_transition(state, block)`. State transitions that trigger an unhandled exception (e.g. a failed `assert` or an out-of-range list access) are considered invalid.
 
-1. State caching, which happens at the start of every slot.
-2. The per-epoch transitions, which happens at the start of the first slot of every epoch.
-3. The per-slot transitions, which happens at every slot.
-4. The per-block transitions, which happens at every block.
-
-Transition section notes:
-* The state caching caches the state root of the previous slot and updates block and state roots records.
-* The per-epoch transitions focus on the [validator](#dfn-validator) registry, including adjusting balances and activating and exiting [validators](#dfn-validator), as well as processing crosslinks and managing block justification/finalization.
-* The per-slot transitions focus on the slot counter.
-* The per-block transitions generally focus on verifying aggregate signatures and saving temporary records relating to the per-block activity in the `BeaconState`.
-
-Beacon blocks that trigger unhandled Python exceptions (e.g. out-of-range list accesses) and failed `assert`s during the state transition are considered invalid.
-
-Note: If there are skipped slots between a block and its parent block, run the steps in the [state-root](#state-caching), [per-epoch](#per-epoch-processing), and [per-slot](#per-slot-processing) sections once for each skipped slot and then once for the slot containing the new block.
-
-### State caching
+```python
+def state_transition(state: BeaconState, block: BeaconBlock) -> BeaconBlock:
+    # Process slots (including those with no blocks) since block
+    process_slots(state, block.slot)
+    # Process block
+    process_block(state, block)
+    # Return post-state
+    return state
```

Comment:
As noted in #1059, `state` is mutable. I think we are mixing how we handle mutable and immutable objects.

For example, writing `process_block(state, block)` implies it's mutable, but then having `return state` is confusing and can be removed since we are expecting that the pyspec user/reader should know `state` is mutable.

Reply:
We are definitely mixing things. Conceptually thinking of `state_transition` as a `(BeaconState, BeaconBlock) -> BeaconState` is cleanest IMO. Will mull.

Reply:
Implemented your great suggestion to add a `try` `except` 🎉